src/Control/Monad/ST.purs

```diff
@@ -49,16 +70,5 @@ foreign import writeSTRef
 -- |
 -- | It may cause problems to apply this function using the `$` operator. The
 -- | recommended approach is to use parentheses instead.
-foreign import runST
-  :: forall a r
-   . (forall h. Eff (st :: ST h | r) a)
-  -> Eff r a
-
--- | A convenience function which combines `runST` with `runPure`, which can be
--- | used when the only required effect is `ST`.
--- |
--- | Note: since this function has a rank-2 type, it may cause problems to apply
--- | this function using the `$` operator. The recommended approach is to use
--- | parentheses instead.
-pureST :: forall a. (forall h. Eff (st :: ST h) a) -> a
-pureST st = runPure (runST st)
+runST :: forall a. (forall h. ST h a) -> a
```

Comment:
I think it might be nice to provide a function to turn an `ST h a` into an `Eff a` as well (no need for the `h` to be polymorphic in that case).

Reply:
> no need for the `h` to be polymorphic in that case

Why's that?

Reply:
Because we can allow the user to modify any heap `h` as long as they are in `IO`.

Reply:
Doesn't that defeat the purpose of `ST`?

Reply:
How so? In Haskell, running an `ST s` computation in `IO` is safe if `s` is `RealWorld` (see `STtoIO` in `base`), which makes sense because otherwise we might end up interleaving modifications in `IO` into some other `ST s` computation (e.g. if the `ST` computation were on another spark). That's my understanding anyway.

But in PS, an `ST` computation has to run to completion in one go, and the same is true of `Eff`, so the same thing can't go wrong. You can't run `Eff` in `ST`, but you should be able to run `ST` in `Eff`.

I'm fine if we want to omit it though, or mark it `unsafe`.

Reply:
Why do `ST a b` and `Eff a b` have to run to completion in PS?

Reply:
Well that is the meaning of values of those types right now. `Eff` is a synchronous function and `ST` is built on top of `Eff`.

Do you mean that we might want to allow `ST` to be asynchronous in future? I think that's a bigger discussion, and I would rather build a separate version of `ST` which was more carefully controlled if we decided to go that route.

Reply:
> Do you mean that we might want to allow ST to be asynchronous in future?

Nope. Just wanted to understand what you meant.

I'm not an expert on how the JS runtime works exactly, but I want to say that I read somewhere that control can change when a function call is invoked. I don't know if that's true or where to find that information again (maybe the ES spec talks about it?). It's still not immediately clear to me that a `ST a b` will run uninterrupted from other computations, including `Eff a b`.

That said, I don't have a compelling argument one way or the other. So in the interest of moving on, let's choose one of the three (or make another suggestion) and I'll add it to this PR:

1. `Control.Monad.ST.stToEff :: forall a e. (forall h. ST h a) -> Eff e a`
1. `Control.Monad.ST.unsafeSTToEff :: forall a e h. ST h a -> Eff e a`
1. `Control.Monad.ST.Unsafe.stToEff :: forall a e h. ST h a -> Eff e a`

Reply:
We can leave it out for now. It was just an idea, but it's probably of limited usefulness anyway.